proposed/runtime-binding.md

```diff
@@ -0,0 +1,304 @@
+
+**User Impact:** When a user runs an application on a machine that does not have `x` installed, but has a later runtime installed, then the app will transparently run on that version. The user does not need to configure their applications to use that later minor version. There is some risk in this scenario that the app will crash due to incompatibilities in the runtime relative to the runtime that the app targets and that it will be difficult to diagnose the cause (incompatibility)and the solution (installing a older runtime). If that app runs successfully, then the user will be happy. The tension between compatibility and deployment convenience is why this scenario is opt-in.
+
+Let's look at an example.
+
+Requested version:
+
+* `2.1.0`
+
+Installed versions:
+
+* `1.1.17`
+* `3.0.0`
+* `3.0.1`
+* `3.1.0`
+* `4.0.0`
+
+In this scenario, the `3.0.1` runtime will be selected.
+
+## Handling Previews
+
+As of 3.0 Preview 2, previews have special binding behavior. In this proposal, previews are proposed to not have special behavior. Apps that depend on preview runtimes can bind to non-preview runtimes and vice-versa. Previews are not considered to be special going forward, with this proposal. This behavior is oriented on enabling applications built with previews to get a chance to run (likely successfully) on a stable version and to make testing stable applications on previews easy.
+
+Assumption: Stable Visual Studio versions only install stable .NET Core versions.
+
+## COM Components
+
+COM-based and other hosted environments have the requirement of loading managed components built for multiple runtime versions, and this set is typically not known a priori.
+
+The application situation, described earlier in the doc, is largely oriented on missing runtime versions and how to handle that. The COM situation is oriented more on "best fit" for a set of components it may load. As a result, COM and other hosts need to artificially roll-forward to the latest acceptable runtime version to enable compatibility for arbitrary components.
+
+We will offer a configuration knobs for hosts that enable artificial roll-forward. These knobs play a similar role to [LockClrVersion](https://docs.microsoft.com/dotnet/framework/unmanaged-api/hosting/lockclrversion-function).
+
+## Configuration Knobs
+
+The host exposes the following configuration knobs that you can use to control binding behavior:
+
+### RollForward
+
+We will introduce a new knob, called `RollForward`, with the following values:
+
+* `None` -- Do not roll forward. Only bind to specified version.
```

Comment:
Does this have a runtimeconfig equivalent today?

Reply:
I was thinking [`applypatches`](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/roll-forward-on-no-candidate-fx.md#applypatches). On that note, is the third example wrong on that page?

@vitek-karas was advocating for not adding `None` to `RollForward`. I find it hard to reason about `--fx-version` behavior in absence of a `None` option.

Reply:
I guess `None` would mean `rollForwardOnNoCandidateFx=0` and `applyPatches=0` in current `runtimeconfig`.
I'm not sure we should "advocate" it by including it in the setting here. Basically we don't want anybody to use that setting unless there's no other way as it disables security patching.

I think that if we decide to have an option for this behavior in `RollForward` it should not be called `None` - mostly because people tend to think about `None` as the safe default - which this is not (neither default nor safe).
If we include it I would call it `DisableAll` or something like that - to make it more explicit and more "scary".